Core/Unit: Fix movement hiccups in water

jackpoz opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

- Add UNIT_FLAG_SWIMMING to Creatures that can enter water when in combat
- Fix creatures with Swim creature_movement_template but without UNIT_FLAG_SWIMMING flag not sticking to the bottom of the sea, i.e. crabs

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #21872

**Tests performed:** (Does it build, tested in-game, etc.)
None

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] Something about charmed units


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

jackpoz committed:
Core/Unit: Allow to define Units which can enter water but cannot swim

Allow to define Units which can enter water but cannot swim, i.e. crabs walking at the bottom of a sea.

jackpoz committed:
Implement the actual change (the previous commit was just renaming preparation)

jackpoz committed:
Add UNIT_FLAG_SWIMMING to creatures that can swim when entering combat

jackpoz committed:
More work for the CanEnterWater()/CanSwim() change

Killyana commented:
I made some tests and the only issues I found is about charmed units (not summoned pets).
If you charm a creature that has "Movement type: Swim: false" then go to the water and brake the charm, it will behave weirdly (it will react passive and keeps following you)
This doesn't happen to creature that cannot walk at the ground, they will just stay in the water, this looks to be the correct way to handle this cases.

So must probably do the same for creature that cannot swim, they will not enter water and remains at the edge.

I changed  "if (IsPet())" to "if (IsPet() && IsSummon())"  but the charmed creature was still able to enter the water

To reproduce:
1. .go c 36874
2. .cast 260
3. go to the water
4. dismiss the creature
5. notice the behaviour

Killyana commented:
Also If you send the controlled unit (.go c 36874) to attack mobs in the water it will not move.
If you .npc info you will see * UNIT_FLAG_SWIMMING (0x8000)  is missing

If you mind control a crab Entry: 2231 it will not get * UNIT_FLAG_SWIMMING (0x8000)  until it starts a combat

Ovahlord commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -2669,6 +2669,13 @@ CreatureMovementData const& Creature::GetMovementTemplate() const
 }
 
+bool Creature::CanEnterWater() const
+{
```
Isn't there CanSwim already? Why not just adding the pet check there instead?

illfated replied:
I don't know the coding involved, but there is a technical difference between creatures being blocked (by their AI) from entering water and the same creatures being able to swim if you spawn them in water.

jackpoz replied:
CanEnterWater() means a Unit can walk into water mesh, i.e. setting the NAV_WWATER filter in MMAPs to allow crossing rivers/lakes and reach enemies in water.

CanSwim() means the Unit is able to lift itself above sea bottom and reach targets floating in the water.

An Unit with CanEnterWater() but without CanSwim() will walk at the bottom of the sea, like a crab for example, and the Z coordinate returned by MMAPs will be GetGroundHeight() and not GetGroundOrWaterLevel().
In current TC crabs sometime get water level,  making them teleport at water level and then instantly sink to the bottom (as they are missing the flags that would allow them to swim)

Ovahlord replied:
The 'teleport' is not properly solved by using the unit flag. It's actually handled by a currently wrong handled spline flag: Walkmode which is actually CanSwim. It has been corrected on master already and I've also backported it for 4.3.4 (and it solves the visual issue you describe) and I am sure this applies to 335 as well.

As for the ground underwater movement I would simply go with Groundlevel by default if CanSwim returns false or the unitflag is not set.

Faq replied:
> As for the ground movement, I would simply go with groundlevel if CanSwim returns false. No need to do additional fancy magic. If a unit is not able to swim it should stick on the ground in all cases. Just move the unit flag and pet check to the CanSwim helper and be done with it and keep the path generator changes to ignore water level of CanSwim returns false.

Sounds plain and simple. If no swim flag, it walks, no matter, there is water or not.

jackpoz replied:
This is already in 335 and that didn't fix the issue

jackpoz commented:
@Killyana a lot of checks are missing in this PR and a lot of cases still need to be handled,  thanks for the tests

VincentVanclef commented on src/server/game/Entities/Unit/Unit.cpp:
```diff
@@ -12647,14 +12647,19 @@ bool Unit::IsFalling() const
-    if (IsPet() && HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_PET_IN_COMBAT))
+    if (HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_PET_IN_COMBAT))
         return true;
     return HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_RENAME | UNIT_FLAG_SWIMMING);
```
why is UNIT_FLAG_RENAME here xD? What does that have to do with swimming?

Shauren replied:
I know it seems stupid but that function was ripped from the client and that was the flag value used there

jackpoz commented:
@Ovahlord did you have another way to solve the same issue or should I continue working on this PR ?

Ovahlord commented:
I think u can just finish this one here.

jackpoz committed:
Merge branch '3.3.5' into fixes/3.3.5-water-hiccups

# Conflicts:
#	src/server/game/Movement/MovementGenerators/HomeMovementGenerator.cpp

jackpoz committed:
Fix charmed creatures not entering water

jackpoz commented:
> If you charm a creature that has "Movement type: Swim: false" then go to the water and brake the charm, it will behave weirdly (it will react passive and keeps following you)

This happens even without going inside any water if I use ".cast 260" . Maybe it's an unrelated bug ?

jackpoz committed:
Half revert previous commit

jackpoz commented:
Turns out PathFinding always adds WATER filter if the creature is in water (according to map/vmap) which makes every creature always go into water anyway

illfated commented:
Hm, so would it be necessary to remove water from quite a lot of specific tiles, to make it work as expected?

jackpoz commented:
more like handle this case in a different way:
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Movement/PathGenerator.cpp#L640-L641

illfated commented:
Interesting, thanks. I see quite a few edge cases in our future. ;-)

jackpoz commented:
This PR contains some useful information to fix the issue (which is in the end related mostly to UNIT_FLAG_SWIMMING flag).
If anyone would like to continue working on it, feel free to do it.

jackpoz committed:
Merge branch '3.3.5' into fixes/3.3.5-water-hiccups

# Conflicts:
#	src/server/game/Entities/Creature/Creature.h
#	src/server/game/Movement/PathGenerator.cpp

jackpoz committed:
Merge branch '3.3.5' into fixes/3.3.5-water-hiccups

jackpoz commented:
> I made some tests and the only issues I found is about charmed units (not summoned pets).
> If you charm a creature that has "Movement type: Swim: false" then go to the water and brake the charm, it will behave weirdly (it will react passive and keeps following you)
> This doesn't happen to creature that cannot walk at the ground, they will just stay in the water, this looks to be the correct way to handle this cases.
> 
> So we must probably do the same for creature that cannot swim, they will not enter water and remains at the edge.
> 
> I changed "if (IsPet())" to "if (IsPet() && IsSummon())" but the charmed creature was still able to enter the water
> 
> To reproduce:
> 
> 1. .go c 36874
> 2. .cast 260
> 3. go to the water
> 4. dismiss the creature
> 5. notice the behaviour

this seems to work now

jackpoz committed:
Always allow Creatures controlled by players to enter water

jackpoz commented:
> Also If you send the controlled unit (.go c 36874) to attack mobs in the water it will not move.
> If you .npc info you will see * UNIT_FLAG_SWIMMING (0x8000) is missing

Fixed this one too with https://github.com/TrinityCore/TrinityCore/pull/24020/commits/322c1d40d5fe1e71b712e39cb8392bb85d73ab76

jackpoz committed:
Add swimming flag when possessing a unit and remove it properly at the end, even if the creature engaged combat before and after.
When adding/removing UNIT_FLAG_SWIMMING manually calling Creature::RefreshSwimmingFlag(true) might be required.

jackpoz commented:
> If you mind control a crab Entry: 2231 it will not get * UNIT_FLAG_SWIMMING (0x8000) until it starts a combat

And fixed this one too

jackpoz commented:
This PR is ready to be tested. If possible please re-test all the cases described in the different comments of this PR. I will host it at https://tcubuntu.northeurope.cloudapp.azure.com/test/ where every new account will have gm seclevel 3 to make things easier to test.

jackpoz committed:
Code cleanup

jackpoz commented:
I cleaned up the code a bit and changed Creature::UpdateMovementFlags() to use CanSwim()

Shauren commented:
creature_movement_template.Swim is now ignored, this is not ok

jackpoz commented:
@Shauren it's checked at https://github.com/TrinityCore/TrinityCore/pull/24020/files#diff-2222cde2ff898034f0ce38bba2d5755cbdeef050c328b2d56599bc6d6664afc1R2713 . If I pull in combat a mob with .Swim false, the mob evades when I enter water and the mob cannot reach me.

gygyti commented:
what about Segacedi?

-> **[https://imgur.com/a/UYbEIJe](https://imgur.com/a/UYbEIJe)**

[https://www.wowhead.com/npc=32664/segacedi](https://www.wowhead.com/npc=32664/segacedi)

Faq commented:
@gygyti make a new issue.

jackpoz commented:
@gygyti have you tried before and after this PR ?

gygyti commented:
before this PR, I did not check Segacedi, there was only a check on 0dee695

gygyti commented:
not sure if this is relevant, but I'll add this video 

https://user-images.githubusercontent.com/77268574/107887339-7c482a80-6f16-11eb-9c9c-68208924b9c1.mp4

jackpoz commented:
any explaination with the video or ?

gygyti commented:
this crab in my opinion should not do such jumps

jackpoz commented:
Are you in combat with it ? Do you have .gm on ? Why is it even following you ?

gygyti commented:
I just hit him to be aggressive, gm is off (it's not a gm account. Regular player )

jackpoz commented:
shouldn't it be red then ? was your character in combat ?

Faq commented:
This goes nowhere, now looks like that is pet. @gygyti just make a new issue where is this all described in details.

gygyti commented:
these are not pets 
entry: 3106, 3104
@Faq I can open a new issue, but everything is described in this thread 

https://user-images.githubusercontent.com/77268574/107984698-8e3ed180-6fd9-11eb-96d8-85ba33651271.mp4


https://user-images.githubusercontent.com/77268574/107984708-9434b280-6fd9-11eb-95a3-11f07acb823d.mp4

jackpoz commented:
@gygyti could you please reproduce that on https://tcubuntu.northeurope.cloudapp.azure.com/ ?

gygyti commented:
wow no, they swim fine on tcubuntu.northeurope.cloudapp.azure.com

gygyti commented:
but why then such a difference?  should i set up worldserver.conf?

illfated commented:
> but why then such a difference? should i set up worldserver.conf?

It could have something to do with the maps/vmaps/mmaps files generated from your game client, and how they affect the creature movement. Troubleshooting your game client map issues could sometimes come down to comparing your output from `.server debug` with one from a better working server.

I also have the disadvantage of owning the enGB client (2006—2010) instead of the enUS one, which in some forums is seen as a guaranteed source of client bugs for gameobjects and more.

TrinityCore rev. c3165dce316d 2021-02-19 18:13:10 +0100 (3.3.5 branch) (Win64, Release, Static)
```sql
TC> server debug
TrinityCore rev. c3165dce316d 2021-02-19 18:13:10 +0100 (3.3.5 branch) (Win64, Release, Static)
Using SSL version: OpenSSL 1.1.1i  8 Dec 2020 (library: OpenSSL 1.1.1i  8 Dec 2020)
Using Boost version: 1.74.0
Using MySQL version: 50732
Using CMake version: 3.19.2
Compiled on: Windows 10.0.19042
Automatic database updates are enabled for the following databases: Auth, Characters, World
Worldserver listening connections on port 8085
Realmlist (Realm Id: 1) configured in port 8085
VMAPs status: Enabled. LineOfSight: 1, getHeight: 1, indoorCheck: 1
MMAPs status: Enabled
maps directory located in ./maps. Total size: 251987687 bytes
vmaps directory located in ./vmaps. Total size: 658130721 bytes
mmaps directory located in ./mmaps. Total size: 1962379500 bytes
Using enUS DBC Locale as default. All available DBC locales: enUS
Using World DB: TDB 335.21021
TC> _
```

gygyti commented:
wow, didn't even think about it. I would like to get a client which is at tcubuntu.northeurope.cloudapp.azure.com 

TrinityCore rev. 8b2fd345e7b0 2020-12-13 12:49:32 +0200 (3.3.5 branch) (Unix, RelWithDebInfo, Static)
Using SSL version: OpenSSL 1.1.1d  10 Sep 2019 (library: OpenSSL 1.1.1d  10 Sep 2019)
Using Boost version: 1.67.0
Using MySQL version: 100327
Using CMake version: 3.13.4
Compiled on: Linux 4.19.0-12-amd64
Automatic database updates are enabled for the following databases: Auth, Characters, World
Worldserver listening connections on port 8085
Realmlist (Realm Id: 1) configured in port 8085
VMAPs status: Enabled. LineOfSight: 1, getHeight: 1, indoorCheck: 1
MMAPs status: Enabled
maps directory located in /data/maps. Total size: 251987687 bytes
vmaps directory located in /data/vmaps. Total size: 658130721 bytes
mmaps directory located in /data/mmaps. Total size: 1962380076 bytes
Using ruRU DBC Locale as default. All available DBC locales: ruRU
Using World DB: TDB 335.21011

illfated commented:
> I would like to get a client which is at tcubuntu.northeurope.cloudapp.azure.com

I think you meant to say that you would like to have a copy the maps/vmaps/mmaps files from the tcubuntu server, but that is understandable. TC does not share binary/compiled files, though, so we may have to get used to having less perfect server files than tcubuntu.northeurope.cloudapp.azure.com . You could try asking in IRC or Discord if someone else has got equally accurate files to share personally with you.

You could also try downloading (torrent) a clean and good enUS game client from a trusted source (read all the available user comments before downloading) and see if that helps you generate more correct dbc/maps/vmaps/mmaps files for your own use. I have not experimented much with that myself, but I sort of know that I should if I want more precise maps than the ones I am using.

Aokromes commented:
notice:

maps directory located in ./maps. Total size: 251987687 bytes
vmaps directory located in ./vmaps. Total size: 658130721 bytes
mmaps directory located in ./mmaps. Total size: 1962379500 bytes

vs

maps directory located in /data/maps. Total size: 251987687 bytes
vmaps directory located in /data/vmaps. Total size: 658130721 bytes
mmaps directory located in /data/mmaps. Total size: 1962380076 bytes

gygyti commented:
interesting difference in size.
I've always been tempted to find the old original 3.3.5 disk and get the data from there. This will definitely be the best.

By the way, the warlock's pets (possibly other mages, hunters) often fall into the ground, and in places like airships (ICC battles on ships, too) they do not obey orders at all. (I'll open a new issue about this)

jackpoz commented:
> rinityCore rev. [8b2fd34](https://github.com/TrinityCore/TrinityCore/commit/8b2fd345e7b0581e74add60db5410e1ae109c5fd) 2020-12-13 12:49:32 +0200 (3.3.5 branch) (Unix, RelWithDebInfo, Static)

This branch was merged on 2021-01-10, update your worldserver
